Moving Nest Functionality

wramstrom opened the pull request:
**Description**

This pull request contains the dycore changes to implement moving nest functionality, as specified in Issue #1100:
https://github.com/ufs-community/ufs-weather-model/issues/1100

This code implements a storm-following moving nest in regional and global configurations.  Nest motion code handles prognostic variables, surface variables, and physics variables.  Nest motion is configured in the namelist, with options for automated storm tracking, prescribed nest motion, and frequency of nest moves.  Automated storm tracking code was ported from HWRF to operate within the FV3 framework.  This code also includes filling SST in a static or moving nest by downscaling from coupled SST in the parent domain.

Moving nest code has been designed to allow prior static global or nested simulations to have identical performance and be bitwise reproducible.  This has been verified with regression testing.  Moving nest code is wrapped with #ifdef MOVING_NEST in existing and new code modules.  The bulk of the new code is in new files in atmos_cubed_sphere/model;  the main files are fv_moving_nest_main.F90, fv_moving_nest.F90, fv_moving_nest_physics.F90, and fv_tracker.F90.   Additional utilities subroutines are added in new files fv_moving_nest_utils.F90 and bounding_box.F90.  The existing module fv_arrays.F90 has significant additions to enable moving physics variables and the automated storm tracking.  The existing module fv_control.F90 has some smaller changes to enable moving nest, and coupling SST in a static or moving nest to an ocean model.

The current code supports a single moving nest, at 3X nest refinement.  Code has been designed to be extensible to multiple moving nests.  

Dependencies:  
This code requires the updated FMS library from PR #840 https://github.com/NOAA-GFDL/FMS/pull/840 to include mpp_shift_nest_domains().

The moving nest code can only be run when the associated PR from the FV3 level calls into the moving nest subroutines added in this PR.  

Fixes #1100 https://github.com/ufs-community/ufs-weather-model/issues/1100

**How Has This Been Tested?**

This code has been tested on Orion and Hera with Intel 18 for a large number of stress cases.  These exercised the moving nest functionality in regional and global configurations.  Most tests were performed with a regional configuration at C512 with a 3X refinement nest that spanned 19 degrees X 19 degrees.  Tests followed storms over the open ocean, through the Caribbean islands, and during continental landfall and subsequent overland passage.  Tests of land interaction are important because a large number of surface fields are specified for land-only points.  In addition, interaction with sharp terrain validates the stability of the model.  Additional tests were performed in the Southern Hemisphere, and crossing the dateline and prime meridian.  Tests focused on the FV3_HAFS_v0_gfdlmp_tedmf physics suite.

Spreadsheet of tests performed is available at: https://docs.google.com/spreadsheets/d/1FP6JReNuCX4hQnEZI1irkooMuw_DiGvXRm1_LxJHhL4/edit#gid=0

The main regression suite now has 5 cases to test moving nest functionality; those config changes will be in the FV3 level pull request for moving nest.  Tests are called: hafs_regional_specified_moving_1nest_atm, hafs_regional_storm_following_1nest_atm, hafs_regional_storm_following_1nest_atm_ocn, hafs_regional_storm_following_1nest_atm_ocn_wav, hafs_global_storm_following_1nest_atm.  

Moving nest development team requests that the commits that make up this PR **NOT** be squashed.

The moving nest capability in the model core was coded by @wramstrom at AOML/HRD with significant collaboration from @BinLiu-NOAA at EMC, and additional collaboration from AOML/HRD (@Xuejin-Zhang, @A-Kyle, Gopal,et al.), EMC (@ChunxiZhang-NOAA, @ZhanZhang-NOAA, @WeiguoWang-NOAA, @JunghoonShin-NOAA, et al.) and GFDL (@lharris4, @bensonr, @TimMarchok-NOAA, @JosephMouallem, et al.).

**Checklist:**

Please check all whether they apply or not
- [X] My code follows the style guidelines of this project
           Follows Fortran style guide, except some lines exceed 120 character length.  
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
           Main fv_moving_nest*.F90 modules have been commented.  Utilities modules are not yet fully commented.     
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] Any dependent changes have been merged and published in downstream modules

BinLiu-NOAA committed:
Merge branch 'dev/emc' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into support/HAFS

BinLiu-NOAA committed:
Remove the hardcoded L64 source code line in tools/external_ic.F90, so that the
model can run with other vertical level configurations.

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/dev/emc' into support/HAFS

wramstrom committed:
Moving Nest Initial Implementation

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/dev/emc' into feature/hafs_moving_nest

wramstrom committed:
Change speed of moving nest to avoid it going off edge of parent tile.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into support/HAFS

BinLiu-NOAA committed:
Merge branch 'support/HAFS' into feature/hafs_moving_nest

wramstrom committed:
Corrected range of value checking for sina and cosa.

wramstrom committed:
Turned down debugging logging.

wramstrom committed:
Added bounds checking so moving nest stays on parent tile.  Only read static latlon grid file once.  Reduced number of variables output by mn_prog_dump_to_netcdf.

wramstrom committed:
Added call to bounds checking so nest stays on parent tile.  Modified to read from static grid files for moving nest only once.  Added crude scaling of nest motion speed based on timestep.

wramstrom committed:
Merged edits to eval_move_nest().

Merge branch 'feature/hafs_moving_nest' of https://github.com/hafs-community/GFDL_atmos_cubed_sphere into feature/hafs_moving_nest

Conflicts:
	driver/fvGFS/atmosphere.F90

wramstrom committed:
Added moving nest code for delz variable.  Improved scaling of automated nest motion.  Removed forced exit at timestep 4800.

wramstrom committed:
Adjusted moving nest edge subroutine.  Will stop motion when 5 points from edge, and block motion for the rest of model run.

wramstrom committed:
Output nest center lat/lon after nest motion.  Reduce logging from permit_move_nest().

wramstrom committed:
Added moving nest support to shift SST/skin temperature.  Added 2D subroutines to moving nest interfaces.  Added performance tracking to init_grid and setup_aligned_nest.

wramstrom committed:
Optimized init_grid/setup_aligned_nest.  Added performance timers.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into feature/hafs_moving_nest
(Bill, Kyle, Xuejin, Bin)

wramstrom committed:
Brought back if-statement that was lost in merge.

wramstrom committed:
Removed obsolete MPP subroutine from use clause.

BinLiu-NOAA committed:
Add merge_import and debug_merge options of atmos_model_nml in driver/fvGFS/atmosphere.F90.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/ch/feature/ic_netcdf' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into support/HAFS

wramstrom committed:
Added nest motion for terrain from static file.

wramstrom committed:
Restructured calls to move nest.

BinLiu-NOAA committed:
*Add the capability of dealing with soil variables for moving nest.
*Update the moving nesting interface.
Note: Coding session among Bill, Xuejin, Kyle, Zhan, Chunxi, Weiguo, Bin

wramstrom committed:
Updates to moving nest for physics variables so code complies and runs.  Added new subroutines for high precision physics variables.

BinLiu-NOAA committed:
*Define the internal vortex tracker needed variables
*Add a fv_diag_tracker subroutine to obtain the tracker needed variables
*Identify and add the call interfaces for the internal vortex tracker in driver/fvGFS/atmosphere.F90.
Note: These are outcome from the coding sessions among Xuejin, Chunxi, Zhan, and Bin.

wramstrom committed:
Working version of mn_phys_dump_to_netcdf

wramstrom committed:
Added soil variables to mn_phys_dump_to_netcdf.  Required changes to pass in IPD data structures.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs_moving_nest' into feature/hafs_vortex_tracking

wramstrom committed:
Added phy_f2d and phy_f3d physics variables to moving nest.  Added land fraction and terrain std to moving nest.

wramstrom committed:
Reinitialize zsurf in fv_diagnostics.F90.  Perform vertical remapping after nest move.  Added nest move for static physics variables read in from file.

wramstrom committed:
Update xlat, xlon, and related arrays in physics after nest move.  Code changes to preserve 64 bit precision of values read in from netCDF, needed for physics variables.

wramstrom committed:
Added many 2D surface physics variables to nest motion.  Corrected call signature for GFS_grid_populate.

wramstrom committed:
Added NSST variables to moving nest functionality.

BinLiu-NOAA committed:
Merge commit '3720aeae54eeed54d4e8274756857542b3e5dccc' into support/HAFS

wramstrom committed:
Removed some logging to speed performance testing.  Initial implementation of physics IFD variable.

wramstrom committed:
Set physics variable IFD from land sea mask (with values flipped).

wramstrom committed:
Handle strong/weak zenith angle albedos facsf facwf by interpolating from parent for moving nest.

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/dev/emc' into support/HAFS

wramstrom committed:
Added terrain_smoother options to namelist, for smoothing edges of terrain in moving nest.

wramstrom committed:
Turn off vertical remapping with nest moves.  Seems not to be needed.

wramstrom committed:
Added 10m winds, total precip, surface temperature to physics nest motion code.

wramstrom committed:
Added comment headers, connected some moving nest main routine variables to namelist.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs_moving_nest' into feature/hafs_vortex_tracking

wramstrom committed:
Removed some logging statements

BinLiu-NOAA committed:
Revert "Add merge_import and debug_merge options of atmos_model_nml in driver/fvGFS/atmosphere.F90."

This reverts commit cd2879def9f4c9b12be483086c2c58ca2b9becb4.

BinLiu-NOAA committed:
*Add model/fv_tracker.F90 containing the FV internal tracker ported from HWRF.
*Add mp_reduce_[min/max][val/loc] in tools/fv_mp_mod.F90 needed by the internal
tracker
*Add internal tracker needed variables and interface in model/fv_arrays.F90,
model/fv_control.F90, and driver/fvGFS/atmosphere.F90.

Contributors: @XuejinZhang-NOAA, @BinLiu-NOAA, @WilliamRamstrom-NOAA,
@ChunxiZhang-NOAA, @ZhanZhang-NOAA

BinLiu-NOAA committed:
Add get_distsq in fv_tracker.F90 to calculate the square of the approximate
distance to the domain center at all points.

BinLiu-NOAA committed:
Internal vortex tracker related clean up in atmosphere.F90 and fv_tracker.F90

BinLiu-NOAA committed:
Fixes for internal tracker related changes so that it can build successfully.

BinLiu-NOAA committed:
Fixes for internal tracker related changes so that it can build successfully.

BinLiu-NOAA committed:
Merge branch 'feature/hafs_vortex_tracking' of https://github.com/hafs-community/GFDL_atmos_cubed_sphere into feature/hafs_vortex_tracking

BinLiu-NOAA committed:
Connect internal tracker with the nest move interface for moving nesting.
Code sprint: @XuejinZhang-NOAA, @WilliamRamstrom-NOAA, @ChunxiZhang-NOAA,
@KyleAhern-NOAA, @ZhanZhang-NOAA, @BinLiu-NOAA

BinLiu-NOAA committed:
Debug FV3 internal vortex tracker and it now can find the storm properly.
Connecting with nest-moving needs further debug though.

BinLiu-NOAA committed:
Connect automatic storm-following and prescribed moving with the nest moving procedure properly.
Contributors: @BinLiu-NOAA, @XuejinZhang-NOAA, @WilliamRamstrom-NOAA.

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/dev/emc' into support/HAFS

BinLiu-NOAA committed:
Merge branch 'feature/hafs_vortex_tracking' into feature/hafs_moving_nest

wramstrom committed:
Commented out call to fms_io_exit() that was needed for moving nest debug outputs.

wramstrom committed:
Added moving nest handling of physics variables - hprime, zorlo, zorlw

wramstrom committed:
Fixed off-by-one error with moving nest terrain data, updated physics version of orography.

wramstrom committed:
Moved temporary moving nest physics grids into fv_atmos_type.  Plus removed some dead code.

wramstrom committed:
Fixed bug in the moving nest physics transfer to fv_arrays.F90

wramstrom committed:
Fixed bug in fv_moving_nest_main.F90 where profiling timers were started repeatedly.

wramstrom committed:
Disabled some debug logging statements.

wramstrom committed:
Switched off detailed moving nest performance profiling.  Moved timestep logging from print statements to call to mpp_error.

wramstrom committed:
Separated moving nest physics code into separate module.  Merged logging code into fv_moving_nest_utils.F90.

wramstrom committed:
Moved physics-related routines into separate file.

BinLiu-NOAA committed:
Remove model/fv_moving_nest_logging.F90, which is no longer needed (Bill).

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into feature/hafs_nesting
Working session among @wramstrom, @BinLiu-NOAA, @Xuejin-Zhang, @ChunxiZhang-NOAA, @A-Kyle.

wramstrom committed:
Improved handling of 32bit/64bit reals as subroutine arguments, to permit 64 bit compiles.

wramstrom committed:
Additional changes to permit 32bit/64bit compiles.

wramstrom committed:
Added handling of new albedo variables to moving nest.  Added handling of lakes to moving nest.

Commit:
Added t-dependent lon-lat (grid_mlont, grid_mlatt) to fv_diagnostics

BinLiu-NOAA committed:
Merge commit '8a3a75903de68ae15527f116087ff2020e86d717' into support/HAFS

wramstrom committed:
Added moving nest functionality for neststruct%i/j stagger.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_nesting

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into feature/hafs_nesting

wramstrom committed:
Added preliminary land masked interpolation for moving nest.

Commit:
fix center-finding arguments in tracker; add debug routine

wramstrom committed:
Regional Moving Nest Enabled. Fixed moving nest for land emissivity and land albedo. Static surface files renamed.

wramstrom committed:
Better handling of longitudes > 180.

wramstrom committed:
Removed logging statements

wramstrom committed:
Moving Nest updates for emis_lnd and emis_wat.

BinLiu-NOAA committed:
Set the default value for surface_dir as INPUT/moving_nest needed by moving
nesting and remove the fix_sfc directory structure inside INPUT/moving_nest.

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/dev/emc' into feature/hafs_nesting

wramstrom committed:
Added patch for nest motion of surface emissivity and albedo.

wramstrom committed:
Merge branch 'feature/hafs_nesting' of https://github.com/hafs-community/GFDL_atmos_cubed_sphere into feature/hafs_nesting

BinLiu-NOAA committed:
Merge branch 'feature/hafs_nesting' of https://github.com/hafs-community/GFDL_atmos_cubed_sphere into feature/hafs_nesting

BinLiu-NOAA committed:
Comment out the dump_surface_physics related debug output in
tools/fv_diagnostics.F90. Other the Cmake build fails due to circular
dependency.

BinLiu-NOAA committed:
Merge remote-tracking branch 'DJ/multiple_output_grids' into feature/hafs_nesting

wramstrom committed:
Only enter moving nest code for moving nest configurations

wramstrom committed:
Code cleanup of moving nest modules

wramstrom committed:
Added doxygen comments and regularized indentation.

BinLiu-NOAA committed:
*Add the capability of downscaling the coupling variables (e.g., SST) from the
parent grid into the nested grid. This is needed since only the atmospheric
parent grid is coupled to other model components currently.
*Update the logic to determine the moving_nesting property.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/DJ/multiple_output_grids' into feature/hafs_nest_cpl

wramstrom committed:
Connected move_nsst to namelist option.

wramstrom committed:
Revised namelist options for move_nsst in moving nest.

wramstrom committed:
Moving nest for physics ifd only if using NSST.

BinLiu-NOAA committed:
Modify atmosphere_domain interface to pass both moving_nest_parent and is_moving_nest up to FV3ATM.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Expand and add the capability to output the grid_mspec (similar to grid_spec)
variables for the moving-nesting configuration.

wramstrom committed:
Added full_zs_filter when nest moves.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs_nest_sync' into feature/hafs_nest_cpl

wramstrom committed:
Upgrade to read in monthly high-resolution albedo information for moving nest.

theurich committed:
Use same real kind for centers and corner... now that ESMF is directly referencing these data arrays.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into feature/hafs_nest_cpl

wramstrom committed:
Handle moving nest interpolation of sfc roughness variables with land mask

BinLiu-NOAA committed:
Merge branch 'feature/hafs_nest_cpl' of https://github.com/hafs-community/GFDL_atmos_cubed_sphere into feature/hafs_nest_cpl

wramstrom committed:
Reapply handling of NSST being turned off for coupled runs.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs_nest_cpl' into feature/hafs_nest_cpl_moving

wramstrom committed:
Added Doxygen documentation and code cleanup.

wramstrom committed:
Merging code cleanup with other changes.

Merge branch 'feature/hafs_nest_cpl' of https://github.com/hafs-community/GFDL_atmos_cubed_sphere into feature/hafs_nest_cpl

Commit:
add zsurf to diag fields for moving nest

wramstrom committed:
Updated formatting of header comments.

laurenchilutti commented:
Do you have a target date for this PR to be merged?

DusanJovic-NOAA commented on model/bounding_box.F90:
```diff
@@ -0,0 +1,175 @@
+    deg2rad = pi / 180.0
+    rad2deg = 1.0 / pi180
+
+    x = in_bbox.is
+    y = in_bbox.js
+    !print '("[INFO] WDR show_bbox ",A8," lats(",I0,",",I0,")=",F10.5," lons(",I0,",",I0,")="F10.5)', tag, x, y, lats(x,y)*rad2deg,  x, y, lons(x,y)*rad2deg
+    x = in_bbox.ie
+    y = in_bbox.js
+    !print '("[INFO] WDR show_bbox ",A8," lats(",I0,",",I0,")=",F10.5," lons(",I0,",",I0,")="F10.5)', tag, x, y, lats(x,y)*rad2deg,  x, y, lons(x,y)*rad2deg
+    x = in_bbox.is
+    y = in_bbox.je
+    !print '("[INFO] WDR show_bbox ",A8," lats(",I0,",",I0,")=",F10.5," lons(",I0,",",I0,")="F10.5)', tag, x, y, lats(x,y)*rad2deg,  x, y, lons(x,y)*rad2deg
+    x = in_bbox.ie
+    y = in_bbox.je
+    !print '("[INFO] WDR show_bbox ",A8," lats(",I0,",",I0,")=",F10.5," lons(",I0,",",I0,")="F10.5)', tag, x, y, lats(x,y)*rad2deg,  x, y, lons(x,y)*rad2deg
```
I am trying to build the ufs-weather-model with GNU compiler. I'm getting this syntax error:
```
/home/dusan/ufs/src/model/FV3/atmos_cubed_sphere/model/bounding_box.F90:136:4:

  136 |     x = in_bbox.is
      |    1
Error: Unclassifiable statement at (1)
make[2]: *** [FV3/atmos_cubed_sphere/CMakeFiles/fv3.dir/build.make:127: FV3/atmos_cubed_sphere/CMakeFiles/fv3.dir/model/bounding_box.F90.o] Error 1
```
I believe all these `x =` and `y =` expressions should be:
```
    x = in_bbox%is
    y = in_bbox%js
```

DusanJovic-NOAA commented on tools/fv_grid_tools.F90:
```diff
@@ -1874,30 +2039,136 @@ subroutine setup_aligned_nest(Atm)
+
+         out_grid = grid_global
+
+         if ( delta_i_c .ne. 0 ) then
+            if (debug_log) print '("[INFO] setup_nest_grid EOSHIFT delta_i_c=",I0," start. npe=",I0)', delta_i_c, this_pe
+            out_grid = eoshift(out_grid, refinement * delta_i_c, 0.0, 1)
+         end if
+
+         if (delta_j_c .ne.  0) then
+            if (debug_log) print '("[INFO] setup_nest_grid EOSHIFT delta_j_c=",I0," start. npe=",I0)', delta_j_c, this_pe
+            out_grid = eoshift(out_grid, refinement * delta_j_c, 0.0, 2)
+         end if
```
Syntax error when using GNU compiler:
```
/home/dusan/ufs/src/model/FV3/atmos_cubed_sphere/tools/fv_grid_tools.F90:2127:65:

 2127 |             out_grid = eoshift(out_grid, refinement * delta_i_c, 0.0, 1)
      |                                                                 1
Error: ‘boundary’ argument of ‘eoshift’ intrinsic at (1) must be the same type and kind as ‘array’
/home/dusan/ufs/src/model/FV3/atmos_cubed_sphere/tools/fv_grid_tools.F90:2132:65:

 2132 |             out_grid = eoshift(out_grid, refinement * delta_j_c, 0.0, 2)
      |                                                                 1
Error: ‘boundary’ argument of ‘eoshift’ intrinsic at (1) must be the same type and kind as ‘array’
make[2]: *** [FV3/atmos_cubed_sphere/CMakeFiles/fv3.dir/build.make:517: FV3/atmos_cubed_sphere/CMakeFiles/fv3.dir/tools/fv_grid_tools.F90.o] Error 1
```
Fix:
```diff
diff --git a/tools/fv_grid_tools.F90 b/tools/fv_grid_tools.F90
index 596a75e..dd32778 100644
--- a/tools/fv_grid_tools.F90
+++ b/tools/fv_grid_tools.F90
@@ -2124,12 +2124,12 @@ contains
 
          if ( delta_i_c .ne. 0 ) then
             if (debug_log) print '("[INFO] setup_nest_grid EOSHIFT delta_i_c=",I0," start. npe=",I0)', delta_i_c, this_pe
-            out_grid = eoshift(out_grid, refinement * delta_i_c, 0.0, 1)
+            out_grid = eoshift(out_grid, refinement * delta_i_c, 0.0_R_GRID, 1)
          end if
 
          if (delta_j_c .ne.  0) then
             if (debug_log) print '("[INFO] setup_nest_grid EOSHIFT delta_j_c=",I0," start. npe=",I0)', delta_j_c, this_pe
-            out_grid = eoshift(out_grid, refinement * delta_j_c, 0.0, 2)
+            out_grid = eoshift(out_grid, refinement * delta_j_c, 0.0_R_GRID, 2)
          end if
 
          shift_p_ind(:,:,1) = shift_p_ind(:,:,1) + delta_i_c
```

wramstrom replied:
I am going to fix this in a different way -- the third argument is a fill value, and defaults to the type-appropriate value of zero for any numeric input arrays.  So I will restructure these calls to remove that argument and explicitly name the final DIM argument.

wramstrom replied:
I have committed the change to remove the BOUNDARY argument, which causes it to use the default value, which is the correct single or double precision value of 0.  This will provide identical functionality as the initial code, but will be portable to various compilers.

wramstrom commented:
> Do you have a target date for this PR to be merged?

We would like to have it ready to merge by end of the month.

wramstrom committed:
Fix syntax error in bounding_box.F90

Corrected syntax error in bounding_box.F90 that was uncovered with gfortran.

lharris4 commented on tools/fv_surf_map.F90:
```diff
@@ -523,7 +523,8 @@ subroutine FV3_zs_filter (bd, isd, ied, jsd, jed, npx, npy, npx_global,  &
       real, intent(inout):: phis(isd:ied,jsd,jed)
-      real, intent(inout):: oro(isd:ied,jsd,jed)
+      !real, intent(inout):: oro(isd:ied,jsd,jed)
+      real, intent(in):: oro(isd:ied,jsd,jed)
```
Why did this need to be changed?

wramstrom replied:
The moving nest code calls FV3_zs_filter() after a nest move, and its call tree has that source orography data as intent(in).  Since the FV3_zs_filter() subroutine does not seem to modify the values of oro, I changed the call to the stricter intent(in).  The internals of FV3_zs_filter() were not changed.

wramstrom committed:
Update Coriolis terms after nest move.

wramstrom committed:
Simplified call to EOSHIFT for compiler portability.  Now omits BOUNDARY argument to use the default value of 0.0 (in single or double precision), instead of explicitly setting it to 0

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/dev/emc' into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Moving-nesting related code clean-ups (Bill, Kyle, Bin).

BinLiu-NOAA committed:
Fix a typo in fill_nested_grid_cpl in driver/fvGFS/atmosphere.F90.

wramstrom committed:
Modularization of moving nest and automated storm tracker.

wramstrom committed:
Separate moving_nest code into its own directory.

BinLiu-NOAA committed:
*fv_tracker.F90 related clean-ups.
*fv_moving_nest_types.F90 clean-ups.

BinLiu-NOAA committed:
Clean up mzsurf from fv_diagnostics.F90 and fv_diagnostics.h (Kyle and Bin).

wramstrom committed:
Made call to EOSHIFT portable to gfortran.

wramstrom committed:
Made call to EOSHIFT portable to gfortran.

Merge branch 'feature/hafs_nest_cpl' of https://github.com/hafs-community/GFDL_atmos_cubed_sphere into feature/hafs_nest_cpl

lharris4 reviewed (changes requested):
This is much improved and what we had wanted to have implemented on this short timescale. I have some minor changes and questions to resolve. Thanks.

wramstrom commented on model/fv_arrays.F90:
```diff
@@ -1251,6 +1254,10 @@ module fv_arrays_mod
 
+
+! For downscaling coupling variables from parent to nested grids
+    real, _ALLOCATABLE :: downcpl2d(:,:) _NULL !< temporary 2d arrary for downscaling coupling variables
```
I am less familiar with the coupling code.  @BinLiu-NOAA can you comment here?

BinLiu-NOAA replied:
@lharris4 This downcpl2d variable is currently used to downscale/pass the SST or sea surface roughness length (z0) variables from the FV3 parent grid to its nest grid. This is because currently the ocean/wave coupling only happens between the FV3atm parent domain and the ocean/wave domains. Thus, we need to pass the SST/z0 fields from the parent to the nest so that the nest domain also gets updated values from the ocean/wave model components. This is currently a temporary solution/workaround. The plan is to enable the coupling/exchange between the FV3atm nested domain and ocean/wave domains directly in future as a better solution.

Alternatively, if you do not want add this downcpl2d variable here. We can move it back inside the subroutine atmosphere_fill_nest_cpl as a temporary array. 

Please feel free to let us know your preference. Thanks!

lharris4 replied:
Hi, Bill, Bin. If it doesn't need to be preserved between calls, then it doesn't need to be in fv_array, and since it is physics-specific it probably doesn't belong here anyway.

Thanks,
Lucas

BinLiu-NOAA replied:
@lharris4, Thanks for the comments/suggestions! I will move it inside the atmosphere_fill_nest_cpl subroutine as a temporary array at this time. And we can revisit this once the direct coupling between the FV3 nested grids and the wave/ocean domains are implemented/enabled. Thanks!

lharris4 replied:
Thanks. We will revisit in a future release.

BinLiu-NOAA replied:
Thanks, @lharris4!

wramstrom commented on tools/fv_diagnostics.F90:
```diff
@@ -4220,17 +4311,33 @@ subroutine range_check_3d(qname, q, is, ie, js, je, n_g, km, pos, q_low, q_hi, b
                do i=is,ie
                   if( q(i,j,k)<q_low .or. q(i,j,k)>q_hi ) then
-                      write(*,998) k,i,j, pos(i,j,1)*rad2deg, pos(i,j,2)*rad2deg, qname, q(i,j,k)
+                      write(*,998) k,i,j, pos(i,j,1)*rad2deg-360.0, pos(i,j,2)*rad2deg, qname, q(i,j,k)
```
It was a convenience change -- most of the test cases were for the Atlantic, and it reported longitudes in the format +270 degrees.  I subtracted 360 to yield a more intuitive value of -90 degrees.   I can remove this change if you prefer.

lharris4 replied:
I would prefer it to be removed. Thank you.

wramstrom replied:
The logging statement has now been reverted to its original format.

wramstrom committed:
Removed commented-out code from fv_control.F90.

wramstrom commented on model/fv_control.F90:
```diff
@@ -161,6 +161,12 @@ module fv_control_mod
    use molecular_diffusion_mod,     only: molecular_diffusion_init, &
                                           read_namelist_molecular_diffusion_nml
 
+#ifdef MOVING_NEST
```
I made some additional edits in response to this comment to remove commented-out code.

lharris4 replied:
OK. Thanks.

wramstrom committed:
Revert logging output in fv_diagnostics.F90 to original format.

BinLiu-NOAA committed:
Rename downcpl2d into parent2nest_2d to reduce confusion. And the
parent2nest_2d variable is to downscale/remap a 2d variable from the parent to
its nest.

lharris4 reviewed (approved):
I am satisfied with the new changes.

bensonr reviewed (commented):
There is an FMS namelist option to have the diagnostic data be dumped on write as opposed to buffered until exit.

lharris4 commented:
Hi, Rusty. I would appreciate it if debugging code were removed.

Thanks,
Lucas

On Fri, Apr 1, 2022 at 11:11 AM Rusty Benson ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> There is an FMS namelist option to have the diagnostic data be dumped on
> write as opposed to buffered until exit.
> ------------------------------
>
> In driver/fvGFS/atmosphere.F90
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179#discussion_r840654110>
> :
>
> > +     ! WDR start code to output moving nest debug information after timestep.
> +     !  This is solely debugging information; not required for moving nest functionality.
> +#ifdef MOVING_NEST
> +
> +     !call date_and_time(TIME=str_time)
> +     !write(message,'("TIMESTEP atmosphere.F90 a_step=",I0," fcst_hr=",F8.2," time=",A12)')  a_step, a_step * dt_atmos / 3600.0, str_time
> +     !call mpp_error(NOTE,message)
> +
> +     !! Re-enable to output buffered NC files early in a run.
> +     !if (tsvar_out .and. a_step .eq. 4800) then
> +     !   call fms_io_exit()   !! Force the output of the buffered NC files
> +     !   print '("[INFO] WDR calling mpp_exit after moving nest atmosphere.F90 npe=",I0)', this_pe
> +     !   call fv_io_exit()
> +     !   call mpp_exit()
> +     !   print '("[INFO] WDR calling STOP after moving nest atmosphere.F90 npe=",I0)', this_pe
> +     !   stop
> +     !endif
> +
> +     !! WDR End code
> +#endif
> +
>
> @lharris4 <https://github.com/lharris4> - did you want this debugging
> code removed?
> ------------------------------
>
> In model/fv_control.F90
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179#discussion_r840662651>
> :
>
> > @@ -161,6 +161,12 @@ module fv_control_mod
>     use molecular_diffusion_mod,     only: molecular_diffusion_init, &
>                                            read_namelist_molecular_diffusion_nml
>
> +#ifdef MOVING_NEST
> +   use fms_io_mod,               only: fms_io_exit
>
> I don't see this use association being used.
> ------------------------------
>
> In model/fv_grid_utils.F90
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179#discussion_r840663705>
> :
>
> > + logical, save       :: first_time = .true.
> + integer, save       :: id_latlon
> +
> +    !if (first_time) then
> +    !   id_latlon     = mpp_clock_id ('latlon2xyz',  flags = clock_flag_default, grain=CLOCK_ROUTINE )
> +    !end if
> +
> +    !call mpp_clock_begin (id_latlon)
> +
>
> This can be removed.
> ------------------------------
>
> In model/fv_grid_utils.F90
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179#discussion_r840664287>
> :
>
> > +    !call mpp_clock_end (id_latlon)
> +
>
> This can be removed.
> ------------------------------
>
> In tools/fv_diagnostics.F90
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179#discussion_r840667305>
> :
>
> > +#ifdef MOVING_NEST
> +!use fv_moving_nest_physics_mod, only : dump_surface_physics
> +!use fms_io_mod,                 only : fms_io_exit
> +!use fv_io_mod,                  only : fv_io_exit
> +!use mpp_mod,                    only : mpp_exit
> +#endif
> +
>
> This can be removed.
> ------------------------------
>
> In tools/fv_diagnostics.F90
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179#discussion_r840668724>
> :
>
> > +#ifdef MOVING_NEST
> +                      !if ( k .eq. km) then
> +!                     print '("[INFO] WDR ",A," RANGE error at lowest level (",I0,",",I0,",",I0,")")', qname, i, j, k
> +!                     call dump_surface_physics(i, j, k)
> +                      !endif
> +#endif
>
> This can be removed.
> ------------------------------
>
> In tools/fv_diagnostics.F90
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179#discussion_r840668932>
> :
>
> > +#ifdef MOVING_NEST
> +!!         !! Dump FMS debugging information
> +!         call fms_io_exit()   !! Force the output of the buffered NC files
> +!         print '("[INFO] WDR calling mpp_exit after moving nest atmosphere.F90 npe=",I0)', mpp_pe()
> +!         call fv_io_exit()
> +!         call mpp_exit()
> +!         print '("[INFO] WDR calling STOP after moving nest atmosphere.F90 npe=",I0)', mpp_pe()
> +!         stop
> +#endif
> +
>
> This can be removed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179#pullrequestreview-929032388>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMUQRVBT7KN5H3MVYMQF243VC4GXTANCNFSM5QNGE35A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

wramstrom committed:
Remove moving nest debugging statements from dycore core.

wramstrom committed:
Remove moving nest debugging statements from dycore.

Merge branch 'feature/hafs_nest_cpl' of https://github.com/hafs-community/GFDL_atmos_cubed_sphere into feature/hafs_nest_cpl

wramstrom commented on driver/fvGFS/atmosphere.F90:
```diff
@@ -689,14 +714,35 @@ subroutine atmosphere_dynamics ( Time )
+
+     !! WDR End code
+#endif
+
```
Thanks for the info on the namelist option.  This debugging code has been removed.

wramstrom commented on model/fv_grid_utils.F90:
```diff
@@ -1766,6 +1776,15 @@ subroutine latlon2xyz(p, e, id)
+    !end if
+
+    !call mpp_clock_begin (id_latlon)
+
```
Code has been removed.

laurenchilutti reviewed (approved):
@wramstrom Can you please create an issue logging the decisions we made in our meeting on March 23rd?  I believe Lucas summarized this in a Google Doc, but it would be helpful to have that information in an issue on Github.

wramstrom commented:
@laurenchilutti I have created the issue here:
[https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/issues/182](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/issues/182)

laurenchilutti commented:
> @laurenchilutti I have created the issue here: #182

Great! Thank you

bensonr commented:
@wramstrom - are you still planning to remove the timers from fv_grid_tools.F90 as discussed?

wramstrom commented:
> @wramstrom - are you still planning to remove the timers from fv_grid_tools.F90 as discussed?

@bensonr I am using the timers in fv_grid_tools.F90 in my optimization work that started this week and runs through May 1.  I can remove them now if it is a significant issue, but would prefer to remove them on the next pass after I have completed performance optimization work.

wramstrom committed:
Moving nest bug fix for nest edge on prime meridian.

BinLiu-NOAA reviewed (approved):
Looks good to me for current phase clean-up. And there will be future PR(s) to further reorganize/optimize/clean up the moving-nesting related code changes.

laurenchilutti commented:
@junwang-noaa this PR is now ready to merge.

laurenchilutti commented:
@wramstrom I see in the description of this PR you would like me to NOT squash commit this.  Why is this?  I want to just note that this is 136 commits.

wramstrom commented:
> @wramstrom I see in the description of this PR you would like me to NOT squash commit this. Why is this? I want to just note that this is 136 commits.

You can squash commit this if needed;  @BinLiu-NOAA showed me some other ways I can examine the evolution of the code if I need it.

DusanJovic-NOAA commented:
@laurenchilutti Please merge this PR. The ufs-weather-model passed full regression tests (https://github.com/ufs-community/ufs-weather-model/pull/1104).
